File: backend-js-ts-express/tests/support/test_token_helper.ts
// backend-js-express/tests/support/test_token_helper.jsの型付き移植。ロジックは変更していない。

import jwt from 'jsonwebtoken';
import crypto from 'node:crypto';

export function makeHmacToken(
  secret: string,
  iss: string,
  aud: string,
  sub: string,
  expOffsetSecs: number,
  alg: jwt.Algorithm = 'HS256',
  azp?: string,
): string {
  const payload: Record<string, unknown> = {
    sub,
    iss,
    aud,
    exp: Math.floor(Date.now() / 1000) + expOffsetSecs,
  };
  if (azp !== undefined) payload.azp = azp;
  return jwt.sign(payload, secret, { algorithm: alg, noTimestamp: true });
}

// "alg":"none"は署名の要らない自己主張トークンになるため、必ず拒否されなければならない。
export function makeAlgNoneToken(iss: string, aud: string, sub: string, expOffsetSecs: number): string {
  const header = { alg: 'none', typ: 'JWT' };
  const payload = {
    sub,
    iss,
    aud,
    exp: Math.floor(Date.now() / 1000) + expOffsetSecs,
  };
  const b64 = (obj: unknown) => Buffer.from(JSON.stringify(obj)).toString('base64url');
  return `${b64(header)}.${b64(payload)}.`;
}

export interface RsaKeyPairPem {
  publicKey: string;
  privateKey: string;
}

export function generateRsaKeyPair(): RsaKeyPairPem {
  return crypto.generateKeyPairSync('rsa', {
    modulusLength: 2048,
    publicKeyEncoding: { type: 'spki', format: 'pem' },
    privateKeyEncoding: { type: 'pkcs8', format: 'pem' },
  });
}

export function makeRsaToken(
  privateKeyPem: string,
  kid: string,
  iss: string,
  aud: string,
  sub: string,
  expOffsetSecs: number,
  alg: jwt.Algorithm = 'RS256',
  azp?: string,
): string {
  const payload: Record<string, unknown> = {
    sub,
    iss,
    aud,
    exp: Math.floor(Date.now() / 1000) + expOffsetSecs,
  };
  if (azp !== undefined) payload.azp = azp;
  return jwt.sign(payload, privateKeyPem, { algorithm: alg, header: { kid, alg }, noTimestamp: true });
}

File: backend-python/tests/unit/support/test_token_helper.py
from __future__ import annotations

import base64
import json
import time

import jwt as pyjwt
from cryptography.hazmat.primitives.asymmetric import rsa


def make_hmac_token(
    secret: str, iss: str, aud: str, sub: str, exp_offset_secs: int, alg: str = "HS256", azp: str | None = None
) -> str:
    payload = {"sub": sub, "iss": iss, "aud": aud, "exp": int(time.time()) + exp_offset_secs}
    if azp is not None:
        payload["azp"] = azp
    return pyjwt.encode(payload, secret, algorithm=alg)


def make_alg_none_token(iss: str, aud: str, sub: str, exp_offset_secs: int) -> str:
    """"alg":"none"は署名の要らない自己主張トークンになるため、必ず拒否されなければならない。
    PyJWTのencode()はセキュリティ上デフォルトでalg=noneを許可しないため、手動で組み立てる
    (backend-java/backend-kotlinの同名テストと同じ意図)
    """
    header = {"alg": "none", "typ": "JWT"}
    payload = {"sub": sub, "iss": iss, "aud": aud, "exp": int(time.time()) + exp_offset_secs}
    header_b64 = _b64url(json.dumps(header).encode())
    payload_b64 = _b64url(json.dumps(payload).encode())
    return f"{header_b64}.{payload_b64}."


def generate_rsa_key_pair() -> rsa.RSAPrivateKey:
    return rsa.generate_private_key(public_exponent=65537, key_size=2048)


def make_rsa_token(
    private_key: rsa.RSAPrivateKey,
    kid: str,
    iss: str,
    aud: str,
    sub: str,
    exp_offset_secs: int,
    alg: str = "RS256",
    azp: str | None = None,
) -> str:
    payload = {"sub": sub, "iss": iss, "aud": aud, "exp": int(time.time()) + exp_offset_secs}
    if azp is not None:
        payload["azp"] = azp
    return pyjwt.encode(payload, private_key, algorithm=alg, headers={"kid": kid})


def rsa_modulus_b64url(public_numbers) -> str:
    return _b64url_uint(public_numbers.n)


def rsa_exponent_b64url(public_numbers) -> str:
    return _b64url_uint(public_numbers.e)


def _b64url_uint(value: int) -> str:
    length = (value.bit_length() + 7) // 8
    return _b64url(value.to_bytes(length, "big"))


def _b64url(data: bytes) -> str:
    return base64.urlsafe_b64encode(data).rstrip(b"=").decode()
